src/NexusMods.DataModel/Repository/Repository.cs

```diff
@@ -0,0 +1,109 @@
+using System.Collections.ObjectModel;
+using System.Reactive.Linq;
+using DynamicData;
+using Microsoft.Extensions.DependencyInjection;
+using NexusMods.Abstractions.MnemonicDB.Attributes;
+using NexusMods.MnemonicDB.Abstractions;
+using NexusMods.MnemonicDB.Abstractions.Models;
+using static System.Reactive.Linq.Observable;
+
+namespace NexusMods.DataModel.Repository;
+
+internal class Repository<TModel> : IRepository<TModel> where TModel : Entity
+{
+    private readonly IConnection _conn;
+    private readonly IAttribute[] _watchedAttributes;
+
+    /// <summary>
+    /// DI constructor.
+    /// </summary>
+    public Repository(IAttribute[] watchedAttributes, IConnection connection)
+    {
+        _conn = connection;
+        _watchedAttributes = watchedAttributes;
+        
+        _cache
+            .Connect()
+            .Bind(out _observable)
+            .Subscribe();
+        
+        _conn.Revisions
+            .SelectMany(db => db.Datoms(db.BasisTxId).Select(datom => (datom, db)))
+            .Where(d => _watchedAttributes.Contains(d.datom.A))
+            .Select(d => (Db: d.db, E: d.datom.E))
+            .StartWith(All.Select(l => (Db: l.Db, E:l.Id)))
+            .Subscribe(dbAndE =>
+            {
+                _cache.Edit(x =>
+                {
+                    x.AddOrUpdate(dbAndE.Db.Get<TModel>(dbAndE.E));
+                });
+            });
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<TModel> All
+    {
+        get
+        {
+            var db = _conn.Db;
+            var items = db.Find(_watchedAttributes[0]);
+            foreach (var attr in _watchedAttributes.Skip(1))
+            {
+                items = items.Intersect(db.Find(attr));
+            }
+
+            return items.Select(id => db.Get<TModel>(id));
+        }
+    }
+    
+    private readonly SourceCache<TModel, EntityId> _cache = new(x => x.Id);
+
+    private readonly ReadOnlyObservableCollection<TModel> _observable;
+
+    /// <inheritdoc />
+    public ReadOnlyObservableCollection<TModel> Observable => _observable;
+
+    /// <inheritdoc />
+    public IObservable<TModel> Revisions(EntityId id)
+    {
+        if (!Exists(id))
+            return Empty<TModel>();
```

Comment:
This feels like a weird implementation detail that might result in hours of wasted debugging. It might very well be possible to want all changes to a yet not-existing model. The observable stream should return no elements either way. Either remove this branch or add a parameter `ignoreMissingModel` with a default value of `false` but definitely document this somewhere.

Reply:
ah good point, I didn't think about the fact that we could have a live stream to a non-existant or badly mapped entity. This is needed because sometimes the UI will ask for `EID:0` as a default value and I don't want that to throw, but there's better ways of getting this effect.